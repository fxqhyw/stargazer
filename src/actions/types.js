export const GET_REPOSITORY = 'GET_REPOSITORY'
export const GET_REPOSITORY_SUCCESSFUL = 'GET_REPOSITORY_SUCCESSFUL'
export const GET_REPOSITORY_FAILURE = 'GET_REPOSITORY_FAILURE'
export const REMOVE_REPOSITORY = 'REMOVE_REPOSITORY'