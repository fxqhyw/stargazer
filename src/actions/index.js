import * as types from './types';

export const getRepository = (fullName) => ({
  type: types.GET_REPOSITORY,
  payload: fullName 
})

export const getRepositorySuccessful = (repository) => ({
  type: types.GET_REPOSITORY_SUCCESSFUL,
  payload: repository
})

export const getRepositoryFailure = (error) => ({
  type: types.GET_REPOSITORY_FAILURE,
  payload: error
})

export const removeRepository = (id) => ({
  type: types.REMOVE_REPOSITORY,
  payload: id
})

export const getRepositoryLanguages = (id) => ({
  type: types.GET_REPOSITORY_LANGUAGES,
  payload: id
})

export const getRepositoryLanguagesSuccessful = ({id, languages}) => ({
  type: types.GET_REPOSITORY_LANGUAGES_SUCCESSFUL,
  payload: {
    id: id,
    languages: languages
  }
})
