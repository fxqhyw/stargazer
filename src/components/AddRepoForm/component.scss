@media all and (min-width: 480px) {
	.add-repo-form {
	  padding: 60px 0;
	}
  
	.add-repo-form form {
	  margin: 0 auto;
	  max-width: 320px;
	}
}
