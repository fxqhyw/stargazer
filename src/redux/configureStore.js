import { createStore, applyMiddleware } from 'redux'
import { composeWithDevTools } from 'redux-devtools-extension/developmentOnly'
import { createLogicMiddleware } from 'redux-logic'
import Reducer from './reducers'
import rootLogic from './logic/'
import axios from 'axios'

const initialState = {
  repositories: {
    data: {},
    error: null
  }
}

const logicDeps = {
    httpClient: axios
}

const logicMiddleware = createLogicMiddleware(rootLogic, logicDeps);


export default function configureStore() {
  return createStore(
    Reducer,
    initialState,
    composeWithDevTools(
      applyMiddleware(logicMiddleware),
    )
  )
}

