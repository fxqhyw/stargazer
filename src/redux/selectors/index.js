import { createSelector } from 'reselect'
import { orderBy } from 'lodash'

const repositories = state => state.repositories

export const repositoryError = createSelector(
  repositories,
  repositories => repositories.error
)

export const repositoriesData = createSelector(
  repositories,
  repositories => orderBy(Object.values(repositories.data), ['stargazers_count'], ['desc'])
)


export const repositoryData = (id) => createSelector(
  repositories,
  repositories => repositories.data[id]
)
