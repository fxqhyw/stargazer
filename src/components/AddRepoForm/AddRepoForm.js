import React, { useState } from 'react';
import { useSelector, useDispatch } from 'react-redux'
import { getRepository } from '../../actions'
import { Form, InputGroup, FormControl, Button } from 'react-bootstrap';
import './AddRepoForm.css'

const AddRepoForm = () => {
  const [repoName, setRepoName] = useState('')
  const error = useSelector(state => state.repositories.error)
  const dispatch = useDispatch()

  const handleSubmit = (e) => {
    e.preventDefault()
    dispatch(getRepository(repoName))
  }

  const handleChange = (e) => {
    setRepoName(e.target.value)  
  }

  return(
    <Form
      className="add-repo-form"
      onSubmit={handleSubmit}>
      <InputGroup className="mb-3">
        <FormControl
          onChange={handleChange}
          placeholder="Enter repository name, eg. facebook/react"
          title="Should be formatted as {owner}/{repo}, eg. facebook/react"
          value={repoName}
          pattern="^([a-zA-z0-9]+\/)([a-zA-z0-9]+)$"
          required
        />
        <InputGroup.Append>
          <Button
            type="submit"
            variant="info">
            Add
          </Button>
        </InputGroup.Append>
      </InputGroup> 
      {error}
    </Form>
  )
}

export default AddRepoForm;
