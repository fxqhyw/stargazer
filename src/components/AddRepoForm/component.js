import React, { useState, useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux'
import { getRepository } from '../../actions'
import { Form, InputGroup, FormControl, Button } from 'react-bootstrap'
import { repositoryError } from '../../redux/selectors'
import './component.scss'

const AddRepoForm = () => {
  const [repoName, setRepoName] = useState('')
  const error = useSelector(repositoryError)
  const dispatch = useDispatch()

  useEffect(() => {
    if (!error) setRepoName('')
  }, [error])

  const handleSubmit = (e) => {
    e.preventDefault()
    dispatch(getRepository(repoName))
  }

  const handleChange = (e) => {
    setRepoName(e.target.value)
  }

  return(
    <Form
      className="add-repo-form"
      onSubmit={handleSubmit}>
      <InputGroup className="mb-3">
        <FormControl
          onChange={handleChange}
          placeholder="Enter repository name, eg. facebook/react"
          title="Should be formatted as {owner}/{repo}, eg. facebook/react"
          value={repoName}
          pattern="^([a-zA-z0-9_.-]+\/)([a-zA-z0-9_.-]+)$"
          required
        />
        <InputGroup.Append>
          <Button
            type="submit"
            variant="info">
            Add
          </Button>
        </InputGroup.Append>
      </InputGroup> 
      <div className="text-danger">
        {error}
      </div>
    </Form>
  )
}

export default AddRepoForm;
