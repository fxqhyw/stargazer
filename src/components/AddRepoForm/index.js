import  AddRepoForm from './AddRepoForm'

export default AddRepoForm
