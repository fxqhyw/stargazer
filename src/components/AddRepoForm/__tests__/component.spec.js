import React from 'react'
import { useDispatch } from 'react-redux'
import { shallow, mount } from 'enzyme'
import AddRepoForm from '../component'
import { getRepository } from '../../../actions'
import { repositoryError } from '../../../redux/selectors'

jest.mock('../../../redux/selectors', () => ({
  repositoryError: jest.fn()
}))

jest.mock('react-redux', () => ({
  useSelector: jest.fn(fn => fn()),
  useDispatch: jest.fn()
}))

describe('AddRepoForm', () => {
  let component;
 
  it('should render correctly', () => {
    repositoryError.mockReturnValue('Repository not found')

    component = shallow(
      <AddRepoForm />
    )
    expect(component).toMatchSnapshot()
  });

  it('should dispatch an action on button click', () => {
    const dispatch = jest.fn()
    useDispatch.mockImplementation(() => dispatch) 

    component = mount(
      <AddRepoForm />
    )
    component.find('input').simulate('change', { target: { value: 'facebook/react' } })
    component.find('form.add-repo-form').simulate('submit')
    expect(dispatch).toHaveBeenCalledTimes(1)
    expect(dispatch).toHaveBeenCalledWith(getRepository('facebook/react'))
  })

  it('should show an error', () => {
    const dispatch = jest.fn()
    useDispatch.mockImplementation(() => dispatch)

    component = mount(
      <AddRepoForm />
    )
    component.find('input').simulate('change', { target: { value: 'ffdsf/rfdfde' } })
    component.find('form.add-repo-form').simulate('submit')
    repositoryError.mockReturnValue('Repository not found')
    expect(component.find('.text-danger').text()).toEqual('Repository not found')
  })

  afterEach(() => {
    component.unmount()
  })
});
