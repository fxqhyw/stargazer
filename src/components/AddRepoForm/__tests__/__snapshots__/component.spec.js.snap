// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddRepoForm should render correctly 1`] = `
<Form
  className="add-repo-form"
  inline={false}
  onSubmit={[Function]}
>
  <InputGroup
    className="mb-3"
  >
    <FormControl
      onChange={[Function]}
      pattern="^([a-zA-z0-9_.-]+\\\\/)([a-zA-z0-9_.-]+)$"
      placeholder="Enter repository name, eg. facebook/react"
      required={true}
      title="Should be formatted as {owner}/{repo}, eg. facebook/react"
      value=""
    />
    <InputGroupAppend>
      <Button
        active={false}
        disabled={false}
        type="submit"
        variant="info"
      >
        Add
      </Button>
    </InputGroupAppend>
  </InputGroup>
  <div
    className="text-danger"
  >
    Repository not found
  </div>
</Form>
`;
